### PR TITLE
added option to switch off validation for ConfigBeanFactory use

### PR DESCRIPTION
**Pull request by dhakehurst**
Useful in its own right I believe, but this also provides a work around for #440

I have beans that I want instantiated by the config file. But I don't want to set every property of the bean. The bean itself sets default values for some properties. I cannot add the '@Optional' annotation to the bean class properties.

switching off the validation allows this.

(I'm using this library for a live project, it is very useful, thanks for all who have contributed towards it.
I need this capability ASAP, and I would rather use an official release over my patched version.)

I have signed the Typesafe Contributor License Agreement online

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -4,7 +4,7 @@
 // Release tags should follow: http://semver.org/
 import scalariform.formatter.preferences._
 
-ThisBuild / git.baseVersion         := "1.3.0"
+ThisBuild / git.baseVersion         := "1.3.4"
```
Comment (havocp) [inline build.sbt]: This is an unrelated change I guess, ideally not in this PR

File: config/src/main/java/com/typesafe/config/ConfigBeanFactory.java [hunk 2]
```diff
@@ -44,6 +44,44 @@
...
+     *
+     * Fields are mapped to config by converting the config key to
+     * camel case.  So the key <code>foo-bar</code> becomes JavaBean
+     * setter <code>setFooBar</code>.
+     *
+     * @since 1.3.4
+     *
+     * @param config source of config information
+     * @param clazz class to be instantiated
+     * @param validate passing false means that the config will not be validated against the bean's implied schema
```
Comment (havocp) [inline config/src/main/java/com/typesafe/config/ConfigBeanFactory.java]: I don't love the boolean param ( long story https://blog.ometer.com/2011/01/20/boolean-parameters-are-wrong/ , tldr `create(config, clazz, false)` doesn't read well).

An alternative would be an Options object like ConfigParseOptions / ConfigResolveOptions, or an enum like `ValidationMode { NONE, AUTOMATIC }` or something.

File: config/src/main/java/com/typesafe/config/ConfigBeanFactory.java [hunk 3]
```diff
@@ -44,6 +44,44 @@
...
+     * The Java class should follow JavaBean conventions. Field types
+     * can be any of the types you can normally get from a {@link
+     * Config}, including <code>java.time.Duration</code> or {@link
+     * ConfigMemorySize}. Fields may also be another JavaBean-style
+     * class.
+     *
+     * Fields are mapped to config by converting the config key to
+     * camel case.  So the key <code>foo-bar</code> becomes JavaBean
+     * setter <code>setFooBar</code>.
+     *
```
Comment (havocp) [inline config/src/main/java/com/typesafe/config/ConfigBeanFactory.java]: I think this kinda needs to explain what happens if we disable validation and it's invalid. I believe the answer is stuff gets left as whatever it was set to originally in the bean (stuff does not get assigned). But full-on type errors (assign an integer to a list or something) maybe still throw?

File: config/src/test/scala/com/typesafe/config/impl/ParseableReaderTest.scala [hunk 4]
```diff
@@ -2,29 +2,29 @@ package com.typesafe.config.impl
...
-      ConfigFactory.parseReader(configInput)
+    @Test
+    def parseIncorrectFormat(): Unit = {
+        val filename = "/test01.properties"
+        val configInput = new InputStreamReader(getClass.getResourceAsStream(filename))
+        val e = intercept[ConfigException.Parse] {
+            ConfigFactory.parseReader(configInput)
+        }
+        assertThat(e.getMessage, containsString("Expecting end of input or a comma, got '^'"))
     }
```
Comment (havocp) [inline config/src/test/scala/com/typesafe/config/impl/ParseableReaderTest.scala]: are there changes in this file or just accidental reformatting?

Comment (dhakehurst) [reply config/src/test/scala/com/typesafe/config/impl/ParseableReaderTest.scala]: I changed nothing here.

File: build.sbt (hunk 1 above)
Comment (dhakehurst) [reply build.sbt]: the current release is 1.3.3, so I don't know why this was set to 1.3.0.
I changed it to 1.3.4, so that my local publication is at a higher version than the release.
(perhaps it should still read 1.3.3 or 1.3.0?)

File: config/src/main/java/com/typesafe/config/ConfigBeanFactory.java (hunk 2 above)
Comment (dhakehurst) [reply config/src/main/java/com/typesafe/config/ConfigBeanFactory.java]: That would be an acceptable alternative. though I am fine with booleans.

Comment (kringol) [reply config/src/main/java/com/typesafe/config/ConfigBeanFactory.java]: I am also looking forward to this feature, since i also need it. 
But i would also think it would be nicer to have more control on the validation. For example to keep the "type" validation for non-null fields, but ignore the optional/null value check, thus allowing null values but validating the type on non-null ones. And similar. 
Currently the boolean is disabling/enabling everything

**Comment by kringol** [comment]
Just added a PR to your own fork with some small changes to match the mechanism of other parts of the library and enable future settings if needed, plus a mini test
https://github.com/dhakehurst/config/pull/1

Thanks :)

> Feel free to add to this pull request or superceed it. I have done all i am able to do at this time.

**Comment by avillev** [comment]
Hello there !! @havocp @dhakehurst 
Do we have new about that change?
Does someone plan to have it merged into the project?

**Comment by dhakehurst** [comment]
I do not currently have more time to work on this

**Comment by kringol** [comment]
I will continue this change on a new PR instead then. 
I really want this moving forward and been waiting for it

-> https://github.com/lightbend/config/pull/616

**Comment by havocp** [comment]
Closing this since it moved to #616
